### DISCUSSION TITLE
Reinitiate the Colorpicker after inserting html with ajax

### DISCUSSION THREAD

**Discussion by SoulFliktioN** [mdbassit/Coloris#135] [Show and tell]
First of all, thanks for this pretty and easy to use script!

I had problems reinitiating the colorpicker after adding some new fields with ajax. So i don't know if there is any better and more correct way but here is my solution:

1. Creating a custom event
`const colorisreinitiate = new Event("COLORISReinitiate");`

2. I changed the DOMReady function to:
```
  function DOMReady(fn, args) {
    args = args !== undefined ? args : [];

    if (document.readyState !== 'loading') {
      fn(...args);
    } else {
      document.addEventListener('DOMContentLoaded', () => {
        fn(...args);
      });
      // added part begin
      document.addEventListener('COLORISReinitiate', () => {
        bindFields(settings.el);
        wrapFields(settings.el);
      });
      // added part end
    }
  }
```

3. Fireing the event in my ajax call
```
$.ajax({
 // ajax options
}}).done(function( data ) {
 document.dispatchEvent(colorisreinitiate);
});
```

**Comment by mdbassit** [comment]
You don't need to do any of that, all you need is to call `Coloris()` again with the same options after your ajax call is done.

```js
$.ajax({
 // ajax options
}}).done(function( data ) {
 Coloris({ /** your options **/ });
});
```

**Reply by SoulFliktioN** [reply]
Thanks for your fast answer! The problem is that the new fields won't get wrapped and therefore there is no thumbnail and the styles aren't applied. 

I made a quick test, just out commented my stuff:
[https://www.isorauschen.at/coloristest.zip](url)

Could it be a browser problem? I am working on Mac OSX newest versions of Google Chrome and Safari.

**Reply by mdbassit** [reply]
Could you please publish the test on your site and send me the link?

**Reply by SoulFliktioN** [reply]
https://www.isorauschen.at/coloris/

**Reply by mdbassit** [reply]
It looks like you set the option `wrap` to `false` when you called `Coloris()` after the ajax call. You should set it to `true` if you want your fields to be wrapped.

**Reply by SoulFliktioN** [reply]
Sorry for this mess and thank you for your answers! I forgot to change it back, but even with set to true it doesn't work. i changed it online.

**Reply by mdbassit** [reply]
Now you have a syntax error in main.js, the second line should be `function initColoris(){`.

**Reply by SoulFliktioN** [reply]
Sorry forgot to upload. Online now.

**Reply by mdbassit** [reply]
You are right, my bad! I wrote this a couple years ago so I forgot that little detail. What you actually need to do is call `Coloris.wrap(selector)`. In this case you are using the default selector `[data-coloris]`, so you just need to call `Coloris.wrap('[data-coloris]')` after the ajax call. 

If you are using a custom selector, for instance a class called "my-color-field", you need to specify that when initializing coloris:

```js
Coloris({
    el: 'my-color-field',
    // the rest of your options
});
```
and after the ajax call:

```js
Coloris.wrap('.my-color-field');
```

**Reply by SoulFliktioN** [reply]
No worries! It is such a great colorpicker. Actually i tried to avoid a custom selector as I would have to change a lot of code. So is there any drawback with calling a new event, binding and wrapping all elements?

**Reply by mdbassit** [reply]
If you don't want to use a custom selector, just use the default `[data-coloris]` one.

The main drawback to your code which adds a custom event is that you have to maintain a modified copy of the color picker.

**Reply by SoulFliktioN** [reply]
Thanks for your effort! And sorry I didn't came up with using the standard selector `[data-coloris]`.

**Reply by mdbassit** [reply]
Always happy to help. I hope that solved your problem.

**Reply by SoulFliktioN** [reply]
Omg yes it just didn't cross my mind to wrap it with the standard selector. I am using this function now and in after my ajax calls simply initColoris(true).

```
function initColoris(wrap){
	Coloris({
		wrap:false,
		theme: 'pill',
		themeMode: 'dark',
		formatToggle: true,
		closeButton: true,
		clearButton: true,
		swatches: [
			'#067bc2',
			'#84bcda',
			'#80e377',
			'#ecc30b',
			'#f37748',
			'#d56062'
		]
	});

	if(wrap){
		Coloris.wrap("[data-coloris]");
	}
}
```

**Reply by mdbassit** [reply]
FYI, `wrap: false` is not effective on the default selector, so you can safely remove it.
